common/network-types/src/frame.rs

```diff
@@ -0,0 +1,534 @@
+/// assert_eq!(bytes.len() / 2, segments.len());
+///
+/// // Create FrameReassembler and feed the segments to it
+/// let (fragmented, reassembled) = FrameReassembler::new(None);
+///
+/// for segment in segments {
+///     fragmented.push_segment(segment).unwrap();
+/// }
+///
+/// drop(fragmented);
+/// pin_mut!(reassembled);
+///
+/// assert_eq!(Some(frame), reassembled.next().await);
+/// # });
+/// ````
+#[derive(Debug)]
+pub struct FrameReassembler<'a> {
+    sequences: DashMap<u32, FrameBuilder<'a>>,
+    sorted_seq_ids: SkipSet<u32>,
+    reassembled: futures::channel::mpsc::UnboundedSender<Frame>,
+    max_age: Option<Duration>,
+}
+
+impl<'a> FrameReassembler<'a> {
+    /// Creates a new frame reassembler and a corresponding stream
+    /// for reassembled [Frames](Frame).
+    pub fn new(max_age: Option<Duration>) -> (Self, impl Stream<Item = Frame>) {
+        let (reassembled, reassembled_recv) = futures::channel::mpsc::unbounded();
+        (
+            Self {
+                sequences: DashMap::new(),
+                sorted_seq_ids: SkipSet::new(),
+                reassembled,
+                max_age,
+            },
+            reassembled_recv,
+        )
+    }
+
+    /// Pushes a new [Segment] for reassembly.
+    pub fn push_segment(&self, segment: Segment<'a>) -> crate::errors::Result<()> {
+        if self.reassembled.is_closed() {
+            return Err(ReassemblerClosed);
+        }
+
+        let ts = current_time().as_unix_timestamp().as_millis() as u64;
+        let seq_id = segment.seq_id();
+
+        match self.sequences.entry(seq_id) {
+            Entry::Occupied(e) => {
+                if e.get().put(segment, ts)? == 0 {
+                    // No more segments missing in this frame, check if this is the earliest frame
+                    if let Some(earliest_seq_id) = self.sorted_seq_ids.front() {
+                        if earliest_seq_id.eq(e.key()) {
+                            let _ = self.reassembled.unbounded_send(e.remove().reassemble()?);
+                            earliest_seq_id.remove();
+                        }
+                    }
+                }
+            }
+            Entry::Vacant(v) => {
+                v.insert(FrameBuilder::new(segment, ts));
+                self.sorted_seq_ids.insert(seq_id);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// If [max_age](FrameReassembler::new) was set during construction, evicts
+    /// leading incomplete frames that are expired at the time this method was called.
+    /// Returns that total number of frames that were evicted.
+    pub fn evict(&self) -> crate::errors::Result<usize> {
+        if self.sequences.is_empty() {
+            self.sorted_seq_ids.clear();
+            return Ok(0);
+        }
+
+        let mut count = 0;
+        if let Some(cutoff) = self
+            .max_age
+            .map(|max_age| current_time().sub(max_age).as_unix_timestamp().as_millis() as u64)
```

Comment:
This clock may not be monotonic, it'd be safer to use the `saturating_sub` to avoid the panic! on invalid operation.

Reply:
This is subtracting `Duration` from `SystemTime`, so not relevant.